blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,823 @@
+how random events can cause extra births or more deaths than expected. This
+kind of equation is known as a stochastic differential equation (SDE).
+Since [DifferentialEquations.jl handles SDEs](http://docs.juliadiffeq.org/latest/tutorials/sde_example.html)
+(and is currently the only library with adaptive stiff and non-stiff SDE integrators),
+these can be handled as a layer in Flux.jl similarly. Here's a neural net layer
+with an SDE:
+
+```julia
+function lotka_volterra_noise(du,u,p,t)
+  du[1] = 0.1u[1]
+  du[2] = 0.1u[2]
+end
+prob = SDEProblem(lotka_volterra,lotka_volterra_noise,[1.0,1.0],(0.0,10.0))
+
+p = param([2.2, 1.0, 2.0, 0.4])
+params = Flux.Params([p])
+function predict_fd_sde()
+  diffeq_fd(p,reduction,101,prob,SOSRI(),saveat=0.1)
+end
+loss_fd_sde() = sum(abs2,x-1 for x in predict_fd_sde())
+loss_fd_sde()
+```
+
+And we can train the neural net to watch it in action and find parameters to make
+the amount of bunnies close to constant:
+
+```julia
+data = Iterators.repeated((), 100)
+opt = ADAM(0.1)
+cb = function ()
+  display(loss_fd_sde())
+  display(plot(solve(remake(prob,p=Flux.data(p)),SOSRI(),saveat=0.1),ylim=(0,6)))
+end
+
+# Display the ODE with the current parameter values.
+cb()
+
+Flux.train!(loss_fd_sde, params, data, opt, cb = cb)
+```
+
+![SDE NN Animation](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)
+
+And we can keep going. There are differential equations
+[which are piecewise constant](http://docs.juliadiffeq.org/latest/tutorials/discrete_stochastic_example.html)
+used in biological simulations, or
+[jump diffusion equations from financial models](http://docs.juliadiffeq.org/latest/tutorials/jump_diffusion.html),
+and the solvers map right over to the Flux.jl neural network frame work through FluxDiffEq.jl
+It is worth noting that FluxDiffEq.jl uses only around ~100 lines of code to pull this all off.
+
+## Implementing the Neural ODE layer in Julia
+
+Let's go all the way back for a second and now implement the neural ODE layer
+in Julia. Remember that this is simply an ODE where the derivative function
+is defined by a neural network itself. To do this, let's first define the
+neural net for the derivative. In Flux, we can define a multilayer perceptron
+with 1 hidden layer and a `tanh` activation function like:
+
+```julia
+dudt = Chain(Dense(2,50,tanh),Dense(50,2))
+```
+
+To define a `neural_ode` layer, we then just need to give it a timespan and use
+the `neural_ode` helper function:
+
+```julia
+tspan = (0.0f0,25.0f0)
+x->neural_ode(dudt,x,tspan,Tsit5(),saveat=0.1)
+```
+
+As a side note, to run this on the GPU, it is sufficient to make the initial
+condition and neural network be on the GPU. This will cause the entire ODE
+solver's internal operations to take place on the GPU without extra data
+transfers in the integration scheme. This looks like:
+
+```julia
+x->neural_ode(gpu(dudt),gpu(x),tspan,BS3(),saveat=0.1)

```

Comment:
Tsit5 doesn't work on the GPU because no one has fixed v1.0 broadcast regressions yet... :( it literally just needs broadcast added to it but we can't because it would slow it down.

Reply:
`# Currently Tsit5 is purposely not GPU compatible due to Julia v1.0 bug` is that good wording?